controllers/trainingjob/trainingjob_controller.go

```diff
@@ -18,411 +18,347 @@ package trainingjob
-	// UPDATE ETCD TO MATCH SM API STATE
-	////////////////////////////////////////////////////////////////////////////////////////////////////
+		if err = r.createTrainingJob(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(err, "Unable to create training job"))
+		}
 
-	if !r.etcdMatchesSmApi(state, describeResponse) {
-		if err = r.updateJobStatus(ctx, log, state, trainingjobv1.TrainingJobStatus{
-			SageMakerTrainingJobName: *state.Spec.TrainingJobName,
-			TrainingJobStatus:        string(trainingJobDescription.TrainingJobStatus),
-			SecondaryStatus:          string(trainingJobDescription.SecondaryStatus),
-			LastCheckTime:            Now(),
-			CloudWatchLogUrl:         cwLogUrl,
-			Additional:               GetOrDefault(trainingJobDescription.FailureReason, ""),
-		}); err != nil {
-			log.Info("Error updating ETCD to sync with SM API state")
-			return RequeueAfterInterval(r.PollInterval, err)
+		if ctx.TrainingJobDescription, err = ctx.SageMakerClient.DescribeTrainingJob(ctx, ctx.TrainingJobName); err != nil {
+			return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(err, "Unable to describe SageMaker training job"))
 		}
-		return RequeueAfterInterval(r.PollInterval, nil)
 	}
 
-	//////////////////////////////////////////////////////////////////////////////////////////////////////////
-	// BASE CASES
-	//////////////////////////////////////////////////////////////////////////////////////////////////////////
+	//TODO: Convert it to tinyurl or even better can we expose CW url via API server proxy UI?
+	ctx.CloudWatchLogURL = "https://" + *ctx.TrainingJob.Spec.Region + ".console.aws.amazon.com/cloudwatch/home?region=" +
+		*ctx.TrainingJob.Spec.Region + "#ctx.LogStream:group=/aws/sagemaker/TrainingJobs;prefix=" +
+		*ctx.TrainingJobDescription.TrainingJobName + ";streamFilter=typeLogStreamPrefix"
 
-	switch trainingJobDescription.TrainingJobStatus {
+	switch ctx.TrainingJobDescription.TrainingJobStatus {
 	case sagemaker.TrainingJobStatusInProgress, sagemaker.TrainingJobStatusStopping:
-		if err = r.updateJobStatus(ctx, log, state, trainingjobv1.TrainingJobStatus{
-			SageMakerTrainingJobName: *state.Spec.TrainingJobName,
-			TrainingJobStatus:        string(trainingJobDescription.TrainingJobStatus),
-			SecondaryStatus:          string(trainingJobDescription.SecondaryStatus),
-			LastCheckTime:            Now(),
-			CloudWatchLogUrl:         cwLogUrl,
-		}); err != nil {
-			log.Info("Error updating ETCD to sync with SM API state")
+		if !controllers.HasDeletionTimestamp(ctx.TrainingJob.ObjectMeta) {
+			if err = r.handleUpdates(ctx); err != nil {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(err, "Unable to update SageMaker trainingjob"))
+			}
+		} else {
+			if _, err := ctx.SageMakerClient.StopTrainingJob(ctx, ctx.TrainingJobName); err != nil {
```

Comment:
I think this will create an unnecessary loop of API calls:
1. TrainingJob is `InProgress`, someone runs `kubectl delete trainingjob x`
2. this line stops the TrainingJob, so TrainingJob enters status `Stopping`
3. `<pollInterval>` seconds later, this switch case sees that the status is `Stopping`. It tries again to `Stop` the SageMaker job.
4. 3 is repeated until job stops (but with exponential backoff instead of `<pollInterval>`).

The operator should only attempt to stop the job if it is not yet stopping or in a terminal state.